src/dmd/astbase.d

```diff
@@ -3633,15 +3636,35 @@ struct ASTBase
         }
     }
 
-    extern (C++) final class TypeClass : Type
+    extern (C++) abstract class TypeAggregate : Type
     {
-        ClassDeclaration sym;
+        AggregateDeclaration asym;
+        alias sym = asym;
+
         AliasThisRec att = AliasThisRec.fwdref;
 
+        extern (D) this(TY ty, AggregateDeclaration asym)
+        {
+            assert(ty == Tstruct || ty == Tclass);
+
+            super(ty);
+            this.asym = asym;
+        }
+
+        override void accept(Visitor v)
+        {
+            v.visit(this);
+        }
+    }
+
+    extern (C++) final class TypeClass : TypeAggregate
+    {
+        extern (D) inout(ClassDeclaration) sym() inout { return cast(typeof(return))asym; }
+        extern (D) void sym(ClassDeclaration asym) { this.asym = asym; }
```

Comment:
I don't see a better way to do this cast. Would be cool to have a way to a static cast in a clean way. I think this will do dynamic cast checks, which are unnecessary, in this case.

Reply:
Right. `asym.isClassDeclaration()` is a virtual call, not sure if that'd be better.

Reply:
Well there is an ugly way: `*cast(ClassDeclaration*) &asym`.